enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+### Non-Goals
+
+- Force an immediate transition for all cloud providers from in-tree support to their out-of-tree counterparts.
+- Exclude in-tree storage provider plugins usage and force transition to CSI storage implementation. 
+- [CSI driver migration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) is out of scope of this work.
+- Support for multiple providers within a single cluster (Though careful operator design may allow for introduction of this in the future).
+
+## Proposal
+
+Our main goal is to start using Cloud Controller Manager in OpenShift 4, and make a seamless transition for all currently supported providers that require the `cloud provider` interface: OpenStack, AWS, Azure, GCP and vSphere.
+
+To maintain the lifecycle of the CCM we want to implement a new cluster operator called `cluster-cloud-controller-manager-operator`.
+This new operator will handle all administrative tasks for the CCM: deployment, restore, upgrade, and so on.
+
+### User Stories
+
+#### Story 1
+
+As a cloud developer, I’d like to improve support for cloud features for OpenShift. I'd like to develop and test new cloud providers which do not ship with default Kubernetes distribution, and require support for external cloud controller manager.
+
+#### Story 2
+
+As a developer I'd like to implement, build and release fixes for cloud provider controllers independently from the Kubernetes core, and assume they will first land upstream and then will be carried over into OpenShift distribution with less effort.
+
+#### Story 3
+
+As an OpenShift developer responsible for the cloud controllers, I want to share more code with upstream Kubernetes in order to ease feature and bug fix contributions in both directions.
+
+#### Story 4
+
+We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.
+
+### Implementation Details
+
+The `cluster-cloud-controller-manager-operator` implementation will be hosted within its own OpenShift [repository](https://github.com/openshift/cluster-cloud-controller-manager-operator). The operator will manage `CCM` provisioning for supported cloud providers.
+
+#### Operator resource management strategy
+
+Every cloud manages its configuration differently, but shares a common overall structure.
+
+1. The operator will be provisioned in `openshift-cloud-controller-manager` namespace.
+2. `CCCMO` will provision operands under the `openshift-cloud-controller` namespace.
+3. The pod running `CCM` controllers in one or more (Azure) containers will be managed by a `Deployment`.
```

Comment:
Where does CCCMO knows which CCM to deploy?

Reply:
Current source-of-truth is the `Infrastructure` resource. CCCMO will consume whatever is/will be the equivalent of it in the future and select a collection of templates to apply.

Reply:
Can you point it out explicitly in this document?